salutations.php

```diff
@@ -248,36 +248,60 @@ function salutation_process_helper($contact_id, $action) {
   $core_greeting_types = array("salutation_email_greeting", "salutation_postal_greeting", "salutation_addressee");
 
   //Salutation Type Field
-  $type_field_id = civicrm_api3('CustomField', 'getvalue', [
-    'return' => "id",
-    'name' => "salutation_type",
-  ]);
+  static $type_field_id = 0;
+  if (!$type_field_id) {
+    $type_field_id = civicrm_api3('CustomField', 'getvalue', [
+      'return' => "id",
+      'name' => "salutation_type",
+    ]);
+  }
   //Salutation Types
-  $salutation_types = civicrm_api3('OptionValue', 'get',[
-    'return' => ["name","value"],
-    'option_group_id.name' => "salutation_type_options",
-    'is_active' => 1,
-  ]);
+  static $salutation_types = [];
+  if (empty($salutation_types)) {
+    $salutation_types = civicrm_api3('OptionValue', 'get',[
+      'return' => ["name","value"],
+      'option_group_id.name' => "salutation_type_options",
+      'is_active' => 1,
+    ]);
+  }
 
-  foreach($salutation_types['values'] as $salutation_type) {
-    $salutation_type_name = $salutation_type['name'];
+  static $greeting_field_id = 0;
+  if (!$greeting_field_id) {
     $greeting_field_id = civicrm_api3('CustomField', 'getvalue', [
       'return' => "id",
       'name' => "salutation_postal_greeting",
     ]);
-    //If creating a new contact, only create the three core greetings by default
-    if ($action == 1 && in_array($salutation_type['value'], $core_greeting_types)) {
-      salutation_create('postal_greeting', $greeting_field_id, $contact_id, $salutation_type['value']);
-    } else {
-      $salutation_option_selected = civicrm_api3('Contact', 'get', [
-        'sequential' => 1,
-        'return' => "custom_$greeting_field_id",
-        'id' => $contact_id,
-        "custom_$type_field_id" => "$salutation_type_name",
-      ]);
-      if ($salutation_option_selected['count'] == 1 &&
-          $salutation_option_selected['values'][0]["custom_$greeting_field_id"] != 4) {
-        salutation_update('postal_greeting', $greeting_field_id, $contact_id, $salutation_option_selected);
+  }
+
+  static $calculated_salutation_id = 0;
+  if (!$calculated_salutation_id) {
+    $calculated_salutation_id = civicrm_api3('CustomField', 'getvalue', [
+      'return' => "id",
+      'name' => "salutation",
+    ]);
+  }
+
+
+  //If creating a new contact, only create the three core greetings by default
+  if ($action == 1) {
+    foreach ($core_greeting_types as $core_greeting_type) {
+      salutation_create('postal_greeting', $greeting_field_id, $contact_id, $core_greeting_type);
+    }
+  } 
+  // Update existing non-customized salutations
+  else {
+    $custom_values = civicrm_api3('CustomValue', 'get', ['entity_id' => $contact_id])['values'];
```

Comment:
Instead of a `Contact.get` I use a `CustomValue.get`.  It delivers the data in a different format, but one API call gets me all the contact's salutation types and greeting type IDs.